content/fewer-braces.md

```diff
@@ -0,0 +1,162 @@
+`<colon> <indent>` and end with `<outdent>`.
+Analogous rules apply for enum bodies, type refinements, and local packages containing nested definitions.
+
+Generally, the possible indentation regions coincide with those regions where braces `{...}` are also legal, no matter whether the braces enclose an expression or a set of definitions. There is so far one exception, though: Arguments to functions can be enclosed in braces but they cannot be simply indented instead. Making indentation always significant for function arguments would be too restrictive and fragile.
+
+To allow such arguments to be written without braces, a variant of the indentation scheme is implemented under language import
+```scala
+import language.experimental.fewerBraces
+```
+This SIP proposes to make this variant the default, so no language import is needed to enable it.
+In this variant, a `<colon>` token is also recognized where function argument would be expected. Examples:
+
+```scala
+times(10):
+  println("ah")
+  println("ha")
+```
+
+or
+
+```scala
+credentials `++`:
+  val file = Path.userHome / ".credentials"
+  if file.exists
+  then Seq(Credentials(file))
+  else Seq()
+```
+
+or
+
+```scala
+xs.map:
+  x =>
+    val y = x - 1
+    y * y
+```
+What's more, a `:` in these settings can also be followed on the same line by the parameter part and arrow of a lambda. So the last example could be compressed to this:
+
+```scala
+xs.map: x =>
+  val y = x - 1
+  y * y
+```
+and the following would also be legal:
+```scala
+xs.foldLeft(0): (x, y) =>
+  x + y
+```
+
+The grammar changes for this variant are as follows.
+
+```
+SimpleExpr       ::=  ...
+                   |  SimpleExpr ColonArgument
+InfixExpr        ::=  ...
+                   |  InfixExpr id ColonArgument
+ColonArgument    ::=  colon [LambdaStart]
+                      indent (CaseClauses | Block) outdent
+LambdaStart      ::=  FunParams (‘=>’ | ‘?=>’)
+                   |  HkTypeParamClause ‘=>’
+```
+### Compatibility
+
+The proposed solution changes the meaning of the following code fragments:
+```scala
+  val x = y:
+    Int
+
+  val y = (xs.map: (Int => Int) =>
+    Int)
+```
+In the first case, we have a type ascription where the type comes after the `:`. In the second case, we have
+a type ascription in parentheses where the ascribing function type is split by a newline. Note that we have not found examples like this in the dotty codebase or in the community build. We verified this by compiling everything with success with `fewerBraces` enabled. So we conclude that incompatibilities like these would be very rare.
+If there would be code using these idioms, it can be rewritten quite simply to avoid the problem. For instance, the following fragments would be legal (among many other possible variations):
+```scala
+  val x = y
+    : Int
+
+  val y = (xs.map: (Int => Int)
+    => Int)
+```
+
+### Other concerns
```

Comment:
I would like to add here a comment that I wrote [there](https://github.com/lampepfl/dotty/pull/15273#issuecomment-1136159846): the proximity with type ascriptions makes code harder to read.

~~~ scala
trait X:
  type Y
  def Y: Y

def f(n: Int): (x: X) => x.Y
f(42): ((x: X) => x.Y) // This is not a lambda but a type ascription
~~~

Even without involving dependent function types, the proximity between brace-less lambdas and function types is quite close:

~~~ scala
// lambda
xs.map: x =>
  x + 1

// type ascription
xs.f: (Int => Int)
~~~

Reply:
`=>` is already heavily overloaded in Scala syntax. In practice, a simple way to visually distinguish a lambda argument from a type ascription will be that the lambda argument ends with a `=>` at the end of the line. Yes, you can fool the logic by putting parentheses around it, as in:
```
  val y = xs.map: ((x: Int) =>
    Int)
```
But I think this will be done very rarely, and we can suggest it should be avoided. (I.e. if you write a multi-line function type, put the `=>`s in front rather at the end of lines.) So then the problem that remains is that _if you intentionally write confusing layout_ and _you read only superficially_ then things can be confusing. But that's really nothing out of the ordinary.